Create pod eviction concept topic (#27739)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/pod-eviction.md

---

> title: Pod Eviction

Reviewer: Is this for the eviction api (the one used when draining a node), or out-of-resource eviction?

Reviewer: This seems to mix details of both.

Reviewer: It is also somewhat strange to have kubelet resource management under a scheduling-related section.

Author: This is meant to include all situations where eviction happens (so from what I understand, kubelet, draining, the eviction API?). The section is Scheduling and Eviction so I put it there - what's your recommendation?

Reviewer: I'm not entirely sure what scheduling and eviction have in common... They seem like two entirely separate topics.

Author: Hmm. What about under Concepts > Cluster Administration instead?

Reviewer: That seems like a reasonable place

Reviewer 2: > I'm not entirely sure what scheduling and eviction have in common... They seem like two entirely separate topics.

Eviction quite often triggers scheduling (of the replacement). I'm in favor of presenting the two concepts alongside each other; in fact, I'd like to include preemption in the mix as well. That way we can contrast preemption with eviction and explain how the two differ - easier if the concepts are presented as siblings.

Also, people who aren't cluster administrators are likely to want to learn about eviction, especially the kind where the kubelet triggers it without prompting.

Reviewer 3: +1 i would move this out of scheduling section and make it clear this is specifically for local kubelet eviction / out-of-resource eviction

Reviewer 2: > I'm not entirely sure what scheduling and eviction have in common... They seem like two entirely separate topics.

Eviction quite often triggers scheduling (of the replacement). I'm in favor of presenting the two concepts alongside each other; in fact, I'd like to include preemption in the mix as well. That way we can contrast preemption with eviction and explain how the two differ - easier if the concepts are presented as siblings.

Also, people who aren't cluster administrators are likely to want to learn about eviction, especially the kind where the kubelet triggers it without prompting.

Author: I'm still a little finicky with github so I'll finish up edits to content before I move it 😝

Author: @sftim makes a good point as well - it made more sense to me under the scheduling section. I will hold off on moving the topic until we decide where it needs a home

Reviewer 4: We had a long discussion in #19081

After seeing this PR, 1/3 types of eviction directly relate to scheduling (currently the PR is not talking about preemption). The others are more about node health, which is under the kubelet domain.

Maybe a better home would be somewhere in https://kubernetes.io/docs/concepts/architecture/

But I know @sftim feels very strongly about this belonging within scheduling. So maybe it's is fine, as long as we change the URL from `scheduling-eviction` to `scheduling-and-eviction`. I thought we had done this already.

Reviewer 2: > change the URL from `scheduling-eviction` to `scheduling-and-eviction`

We usually omit short words like “and” from URLs - for example, “Services And Networking” is represented in the URL as `services-networking`.

content_type: concept
weight: 50
---

{{< glossary_definition term_id="eviction" length="all" >}}

During an eviction, the kubelet sets the `PodPhase` for the selected pods to `Failed`.

> This gracefully terminates the pods. 

Reviewer: Note that it does not respect the pod's terminationGracePeriodSeconds, and uses a 0s grace-period.

If the pods are managed by a workload resource (such as {{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}}
or {{< glossary_tooltip text="Deployment" term_id="deployment" >}}) that

> replaces failed pods, the control plane creates new pods in place of the evicted pods.

Author: What about "the scheduler creates..."?

> replaces failed pods, the control plane creates new pods in place of the evicted pods.

Author: Replaced with "the control plane or `kube-controller-manager`"

{{<note>}}

> The kubelet attempts to [reclaim node-level resources](#reclaim-node-resources)

Reviewer: might be helpful to give an example, like removing unused images when under disk pressure.

before it terminates end-user pods.
{{</note>}}

> When evicting pods, the kubelet respects your configured [PodDisruptionBudget](/docs/tasks/run-application/configure-pdb/).

Reviewer: This is false.  It does not respect PDBs in out-of-resource scenarios.  When using the [eviction API](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api), PDB is respected.  But it is not respected during kubelet evictions.

Author: So there's two main ways in which eviction is triggered - out-of-resource scenarios and the eviction API/draining a node? Would those fall under automated and manual eviction respectively?

Reviewer: Yes, but i'm trying to stress that "eviction" isn't one thing in kubernetes.  Its two completely separate things that were unfortunately given the same name.  But they share no characteristics except the name, and the fact that pods stop running where they are currently running.

Author: Okay. This topic will be the perfect spot to explain the distinction between the two types then, and to stress that they shouldn't and can't be interchangeable. I'll add content that tries to do that and request a review when I do!

> You can control which pods are evicted last by setting the [Quality of Service class](/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed)

Reviewer: This is incorrect.  QoS has no impact on eviction ordering.  See https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/#evicting-end-user-pods for what is taken into account.

Author: Thanks for the comments! 

The [Priority and Preemption page](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#interactions-of-pod-priority-and-qos) says that the only time Priority and QoS are both considered is during out-of-resource eviction. 

Also, the out of resource handling section has `BestEffort` and `Burstable` going before `Guaranteed`. Isn't that considering the QoS?

Reviewer: Yeah, we have been discussing that recently, as you aren't the only one confused by it.  We don't actually take QoS into account, but since QoS is based on setting requests, and requests relative to usage, you can reason from QoS to rough ordering of evictions.  For example, a Guaranteed pod is still going to be evicted last simply because it is impossible for its usage to exceed requests, which is the first criterion we rank by.  Suggestions would be welcome, but we have been discussing removing all references to QoS here.

Reviewer: We should update the Priority and Preemption page.  I wasn't aware of that section.  Note that the text after mentioning QoS where it talks about the algorithm doesn't mention QoS at all...  It is just the headline that is wrong there.

Reviewer 2: @dashpole please feel very welcome to file an issue about any incorrect content you see!

Author: I created #27753 and added a PR for it. Accidentally didn't make the PR a draft, sorry.

on those pods to `Guaranteed`.

The kubelet uses various parameters to make eviction decisions, like the following:

  * Eviction signals
  * Eviction thresholds
  * Monitoring intervals
  * Node conditions

## Triggering eviction

You can configure pod eviction to happen automatically, triggered by a controller. 

> You can also request pod eviction manually using the `kubectl drain` command

Reviewer: This is different from a kubelet eviction.

Author: To clarify, this topic is for all types of eviction. The [Safely Drain a Node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#use-kubectl-drain-to-remove-a-node-from-service) page states that `kubectl drain` evicts pods in a node, so I added that here as one of the possible methods.

Reviewer: Nearly all of this page does not apply to draining a node.  The eviction API does graceful deletion (which eviction is not), and _does_ respect pod disruption budgets, for example.  Your pod also won't be set to phase == Failed during a drain.

signal. Each signal supports either a percentage or a literal value. Kubelet 
calculates the percentage value relative to the total capacity associated with
the signal. 

The value for `memory.available` is derived from the cgroupfs instead of tools
like `free -m`. This is important because `free -m` does not work in a
container, and if users use the [node
allocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable) feature, out of resource decisions
are made local to the end user Pod part of the cgroup hierarchy as well as the
root node. This [script](/examples/admin/resource/memory-available.sh)
reproduces the same set of steps that the `kubelet` performs to calculate
`memory.available`. The `kubelet` excludes inactive_file (i.e. # of bytes of
file-backed memory on inactive LRU list) from its calculation as it assumes that
memory is reclaimable under pressure.

The kubelet supports the following filesystem partitions:

1. `nodefs`: The node's root filesystem, used for local disk volumes, emptyDir,
   log storage, and more.
1. `imagefs`: An optional filesystem that container runtimes use to store container
   images and container writable layers.

> Kubelet auto-discovers these filesystems using cAdvisor and ignores other

Reviewer: cAdvisor is an implementation detail that only applies to linux.  I'd suggest not mentioning it here.

Author: Done.

holding for the defined grace periods. This causes the reported node condition
to constantly switch between `true` and `false`, leading to bad eviction decisions.

To protect against oscillation, you can use the `eviction-pressure-transition-period`
flag, which controls how long the kubelet must wait before transitioning a node
condition to a different state.

## Reclaiming node level resources {#reclaim-node-resources}

The kubelet tries to reclaim node-level resources before it evicts end-user pods.

When a `DiskPressure` node condition is reported, the kubelet reclaims node-level
resources based on the filesystems on the node. 

### With `imagefs`

If the node has a dedicated `imagefs` filesystem for container runtimes to use,
the kubelet does the following:

  * If the `nodefs` filesystem meets the eviction threshlds, the kubelet deletes
    dead pods and containers. 
  * If the `imagefs` filesystem meets the eviction thresholds, the kubelet
    deletes all unused images. 

### Without `imagefs`

If the node only has a `nodefs` filesystem that meets eviction thresholds,
the kubelet frees up disk space in the following order:

> 1. Delete dead pods and containers

Reviewer 3: "Garbage collect" might be more appropriate here

Author: Just for 1. or for both?

Author: ping @bobbypage

1. Delete unused images

## Evicting end-user pods

If the kubelet's attempts to reclaim node-level resources don't bring the eviction
signal below the threshold, the kubelet begins to evict end-user pods. 

The kubelet uses the following parameters to determine pod eviction order:

> 1. Pod resource usage relative to defined requests

Reviewer: This reads as being the same as number 3.  Suggestion: "If pod resource usage exceeds defined requests or not"

Author: Out of curiosity, what's the difference between 1 and 3?

Reviewer: 1. Requests < Usage ?  True or false
2. ...
3. Usage - Requests

Author: OH. Okay, point 3 should be clearer - by how much does usage exceed requests?

1. [Pod priority](/docs/concepts/configuration/pod-priority-preemption/)
1. Resource consumption relative to scheduling requests

> 1. Pod [QoS class](/docs/tasks/configure-pod-container/quality-service-pod/)

Reviewer: This isn't correct.  It is just the first 3 items in this list.

> As a result, kubelet ranks and evicts pods in the following order:

Reviewer: This only applies to CPU and memory, as QoS doesn't take ephemeral storage requests into account.

Author: So something like "Note: the QoS class ranking does not apply to ephemeral storage requests like `inodes` and `PIDs`."?

Reviewer: I'd suggest framing this section differently, if we keep it at all.  The QoS descriptions are meant to help users reason about behavior, rather than completely describe the algorithm.
```
For example, if a node is under MemoryPressure, it would evict pods in the following order:

1. BestEffort or Burstable pods are the only pods where memory usage is allowed to exceed requests.  Among pods where usage is greater than requests, those with the lowest Priority, and then those using the most in excess of their requests are evicted first.
2. `Guaranteed` pods may not exceed their requests, since their requests and limits are equal.  Burstable pods may also not be exceeding their requests when MemoryPressure is triggered.  These pods are the last to be evicted since they are operating within their requested resources.

Note: QoS is not actually used during ranking, but can help illustrate which pods are most likely to be evicted.  QoS does not take EphemeralStorage requests into account, so the above scenario won't apply if the node is under DiskPressure, for example.
```

Author: I like that suggestion a lot, especially based on your other comment about potentially removing QoS at some point. I'll rework.

1. `BestEffort` or `Burstable` pods where the usage exceeds requests. These pods
   are evicted based on their Priority and then by how much their usage level
   exceeds the request.
1. `Guaranteed` pods and `Burstable` pods where the usage is less than requests
   are evicted last, based on their Priority.

`Guaranteed` pods are guaranteed only when requests and limits are specified for
all the containers and they are equal. These pods will never be evicted because
of another pod's resource consumption. If a system daemon (such as `kubelet`,
`docker`, and `journald`) is consuming more resources than were reserved via 
`system-reserved` or `kube-reserved` allocations, and the node only has
`Guaranteed` or `Burstable` pods using less resources than requests left on it,
then the kubelet must choose to evict one of these pods to preserve node stability

> and to limit the impact of resource starvation on other Pods. In this case, it

Reviewer 3: I'm not sure regarding the `system daemon` section. 

@dashpole to keep me honest, but as far as I'm aware kubelet eviction for memory is only looking at the kubepods cgroup `statsapi.SystemContainerPods`: https://github.com/kubernetes/kubernetes/blob/fbc93bd34c51ba81aebf789da67bfa497fa9225a/pkg/kubelet/eviction/helpers.go#L672 

So I don't think memory usage outside the pod cgroup (i.e. those by system daemons) would influence kubelet memory eviction.

Author: Thanks! I'll wait for confirmation  and then add it.

Reviewer: There are two memory eviction thresholds.

1. Sum(Pods) < Allocatable
2. All usage (including pods, system daemons) < Capacity - eviction threshold

System daemon usage isn't limited by kube-reserved in most setups, and comes into play for the second threshold.

Author: Leaving as-is for now.

Reviewer 3: ah got it, thanks @dashpole for clarification. For #2 that checks is from here looks like: https://github.com/kubernetes/kubernetes/blob/fbc93bd34c51ba81aebf789da67bfa497fa9225a/pkg/kubelet/eviction/helpers.go#L666-L670 (i.e. `summary.Node.Memory`)

will choose to evict pods of lowest Priority first.

> When the kubelet evicts pods to reclaim disk resources, it sorts pods by QoS and

Reviewer: This isn't correct.  It is by the same algorithm as CPU and memory.  Note that there are no requests for inodes or PIDs, so priority ends up being the first criterion.

Author: So for disk resources, it uses the same logic as for CPU and memory with the exception of using the QoS class? And then for `inodes` and `PIDs`, only the Priority is used to rank the pods?

Reviewer: For disk, inodes, and PIDs, it uses the same logic as for memory.  (There isn't CPU eviction).  It does _not_ ever take QoS into account.  There are no requests for inodes or PIDs, so the algorithm:

1. Usage > request ?
2. Priority
3. Usage - request

Becomes 

1. Usage > 0 ? (always true)
2. Priority
3. Usage - 0 = Usage

For inodes and PIDs.

`inode` starvation, it uses the QoS to make eviction decisions. The kubelet sorts
pods differently based on whether the node has a dedicated `imagefs` filesystem:

### With `imagefs`

If `nodefs` is triggering evictions, the kubelet sorts pods based on `nodefs`
usage (`local volumes + logs of all containers`).

If `imagefs` is triggering evictions, the kubelet sorts pods based on the
writable layer usage of all containers.

### Without `imagefs`

If `nodefs` is triggering evictions, the kubelet sorts pods based on their total
disk usage (`local volumes + logs & writable layer of all containers`)

## Minimum eviction reclaim

In some cases, pod eviction only reclaims a small amount of the starved resource.
This can lead to the kubelet repeatedly hitting the configured eviction thresholds
and triggering multiple evictions. 

You can use the `eviction-minimum-reclaim` flag to configure a minimum reclaim
amount for each resource. When the kubelet notices that a resource is starved,
it continues to reclaim that resource until it reclaims the quantity you specify. 

For example, the following configuration sets minimum reclaim amounts: 

```

> --eviction-hard=memory.available<500Mi,nodefs.available<1Gi,imagefs.available<100Gi

Reviewer 3: we likely want to point to the kubelet config variant of this config rather than the flags.

Author: Do you mean the YAML config? Where would I get my hands on that? ✋🏽

Reviewer 3: yes, as defined here https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/

for example:

```
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
evictionHard:
    memory.available:  "200Mi"
```

full list of options are defined here: https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/

--eviction-minimum-reclaim="memory.available=0Mi,nodefs.available=500Mi,imagefs.available=2Gi"`
```

In this example, if the `nodefs.available` signal meets the eviction threshold,
the kubelet reclaims the resource until the signal reaches the threshold of `1Gi`,
and then continues to reclaim the minimum amount of `500Mi` it until the signal
reaches `1.5Gi`. 

Similarly, the kubelet reclaims the `imagefs` resource until the `imagefs.available`
signal reaches `102Gi`. 

The default `eviction-minimum-reclaim` is `0` for all resources.

## Node out of memory behavior

> If the node experiences a system out of memory (OOM) event prior to the `kubelet`

Reviewer 3: I would remove "system" and just say out of memory (OOM).

There can be either cgroup or system OOMs. Most likely case is a cgroup OOM, e.g. a burstable pod which exceeds memory limit of node allocatable (top level pod cgroup) which then triggers a cgroup OOM.

```
--eviction-hard=memory.available<500Mi
--system-reserved=memory=1.5Gi
```

In this configuration, the `--system-reserved` flag reserves `1.5Gi` of memory
for the system, which is `10% of the total memory + the eviction threshold amount`. 

The node can reach the eviction threshold if a pod is using more than its request,
or if the system is using more than `1Gi` of memory, which makes the `memory.available`
signal fall below `500Mi` and triggers the threshold. 

### DaemonSet

Pod Priority is a major factor in making eviction decisions. If you do not want
the kubelet to evict pods that belong to a `DaemonSet`, give those pods a high
enough `priorityClass` in the pod spec. You can also use a lower `priorityClass`
or the default to only allow `DaemonSet` pods to run when there are enough 
resources.

## Known issues

The following sections describe known issues related to out of resource handling.

### kubelet may not observe memory pressure right away

The `kubelet` currently polls `cAdvisor` to collect memory usage stats at a regular interval. If memory usage
increases within that window rapidly, the `kubelet` may not observe `MemoryPressure` fast enough, and the `OOMKiller`

> will still be invoked. We intend to integrate with the `memcg` notification API in a future release to reduce this

Reviewer: memcg integration is done, but is not enabled by default, IIRC.  --kernel-memcg-notification is the flag to enable it on the kubelet.

Author: Is the workaround on line 357 still applicable?

Reviewer: Using Node allocatable is probably a better solution. (by setting kube-reserved and system-reserved)

latency, and instead have the kernel tell us when a threshold has been crossed immediately.

If you are not trying to achieve extreme utilization, but a sensible measure of overcommit, a viable workaround for
this issue is to set eviction thresholds at approximately 75% capacity. This increases the ability of this feature
to prevent system OOMs, and promote eviction of workloads so cluster state can rebalance.

### kubelet may evict more Pods than needed

> The Pod eviction may evict more Pods than needed due to stats collection timing gap. This can be mitigated by adding

Reviewer: This is done

---
title: Pod Eviction
content_type: concept
weight: 50
---

{{< glossary_definition term_id="eviction" length="all" >}}

> During an eviction, the kubelet sets the `PodPhase` for the selected pods to `Failed`.
> This gracefully terminates the pods. If the pods are managed by a Deployment, the
> Deployment creates new pods for Kubernetes to schedule. 

Reviewer 2: ```suggestion
During an eviction, the kubelet sets the `PodPhase` for the selected pods to `Failed`.
This gracefully terminates the pods.

If the pods are managed by a workload resource (such as
{{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}}
or {{< glossary_tooltip text="Deployment" term_id="deployment" >}}) that
replaces failed pods, the control plane creates new pods in place of the pod(s)
that were evicted.
```
?

Author: Done

| Eviction Signal      | Description                                                                           |
|----------------------|---------------------------------------------------------------------------------------|
| `memory.available`   | `memory.available` := `node.status.capacity[memory]` - `node.stats.memory.workingSet` |
| `nodefs.available`   | `nodefs.available` := `node.stats.fs.available`                                       |
| `nodefs.inodesFree`  | `nodefs.inodesFree` := `node.stats.fs.inodesFree`                                     |
| `imagefs.available`  | `imagefs.available` := `node.stats.runtime.imagefs.available`                         |
| `imagefs.inodesFree` | `imagefs.inodesFree` := `node.stats.runtime.imagefs.inodesFree`                       |
| `pid.available`      | `pid.available` := `node.stats.rlimit.maxpid` - `node.stats.rlimit.curproc`           |

In this table, the `Description` column shows how kubelet gets the value of the
signal. Each signal supports either a percentage or a literal value. Kubelet 
calculates the percentage value relative to the total capacity associated with
the signal. 

> The value for `memory.available` is derived from the cgroupfs instead of tools

Author: I'm not 100% sure what this means or if it is accurate. (line 59-68)

like `free -m`. This is important because `free -m` does not work in a
container, and if users use the [node
allocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable) feature, out of resource decisions
are made local to the end user Pod part of the cgroup hierarchy as well as the
root node. This [script](/examples/admin/resource/memory-available.sh)
reproduces the same set of steps that the `kubelet` performs to calculate
`memory.available`. The `kubelet` excludes inactive_file (i.e. # of bytes of
file-backed memory on inactive LRU list) from its calculation as it assumes that
memory is reclaimable under pressure.

The kubelet supports the following filesystems:

1. `nodefs`: Volumes, daemon logs, etc.

> 1. `imagefs`: Container runtimes use `imagefs` to store container images and

Reviewer 2: What are nodefs and imagefs? (For example, would I find these names in `/proc/mounts`?)

(I think these are symbolic names used within the kubelet and its configuration, and not used elsewhere). We should clarify what these names mean.

Author: `nodefs` and `imagefs` do seem like the symbolic names for the node's root filesystem (nodefs) and the runtime filesystem (imagefs), which in Docker is apparently "thin pool". 

I reworded these points to: 

1. `nodefs`: The node's root filesystem, used for local disk volumes, emptyDir, daemon logs, and more.
1. `imagefs`: An optional filesystem that container runtimes use to store images and writable layers.

Reviewer 2: Further nit: when Kubernetes people say “node's root filesystem” they mean “filesystem containing `/var/lib/kubelet`” which may not be the filesystem that contains `/`.
I'm not sure if there's a good name for it though. “Node's root filesystem” is nice and short, just sometimes wrong.

Author: @dashpole or @bobbypage can you confirm?

Reviewer: The above are all correct.

Author: changed to 

1. `nodefs`: The node's root filesystem, used for local disk volumes, emptyDir,
   log storage, and more. For example, `nodefs` can provide `/` or `/var/lib/kubelet`. 

Does that cover potential confusion?

Reviewer 2: Strictly speaking _nodefs_ refers to the filesystem containing the kubelet's main directory (that is, `/var/lib/kubelet`). If we can avoid writing `/` or mentioning “root” that avoids a possible point of confusion.

Author: Updated to 1. `nodefs`: The node's main filesystem, used for local disk volumes, emptyDir,
   log storage, and more. For example, `nodefs` contains `/var/lib/kubelet/`.

   container writable layers. This is optional.

Kubelet auto-discovers these filesystems using cAdvisor and ignores other
filesystems. Kubelet does not support other configurations. For example, do not
store volumes and logs in a dedicated filesystem.

> {{<note>}}

Author: Is this still a "future release" feature or has garbage collection been deprecated?

When a `DiskPressure` node condition is reported, the kubelet reclaims node-level
resources based on the filesystems on the node. 

### With `imagefs`

If the node has a dedicated `imagefs` filesystem for container runtimes to use,
the kubelet does the following:

  * If the `nodefs` filesystem meets the eviction threshlds, the kubelet deletes
    dead pods and containers. 
  * If the `imagefs` filesystem meets the eviction thresholds, the kubelet
    deletes all unused images. 

### Without `imagefs`

If the node only has a `nodefs` filesystem that meets eviction thresholds,
the kubelet frees up disk space in the following order:

1. Delete dead pods and containers
1. Delete unused images

## Evicting end-user pods

If the kubelet's attempts to reclaim node-level resources don't bring the eviction
signal below the threshold, the kubelet begins to evict end-user pods. 

The kubelet uses the following parameters to determine pod eviction order:

1. Pod resource usage relative to defined requests
1. [Pod priority](/docs/concepts/configuration/pod-priority-preemption/)
1. Resource consumption relative to scheduling requests

> 1. Pod [QoS](/docs/tasks/configure-pod-container/quality-service-pod/)

Reviewer 2: ```suggestion
1. Pod [QoS class](/docs/tasks/configure-pod-container/quality-service-pod/)
```
?

   are evicted based on their Priority and then by how much their usage level
   exceeds the request.
1. `Guaranteed` pods and `Burstable` pods where the usage is less than requests
   are evicted last, based on their Priority.

`Guaranteed` pods are guaranteed only when requests and limits are specified for
all the containers and they are equal. These pods will never be evicted because
of another pod's resource consumption. If a system daemon (such as `kubelet`,
`docker`, and `journald`) is consuming more resources than were reserved via 
`system-reserved` or `kube-reserved` allocations, and the node only has
`Guaranteed` or `Burstable` pods using less resources than requests left on it,
then the kubelet must choose to evict one of these pods to preserve node stability
and to limit the impact of resource starvation on other Pods. In this case, it
will choose to evict pods of lowest Priority first.

When the kubelet evicts pods to reclaim disk resources, it sorts pods by QoS and
terminates pods using the largest amount of disk. When the kubelet responds to
`inode` starvation, it uses the QoS to make eviction decisions. The kubelet sorts

> pods differently based on whether the node has a dedicated `imagefs` filesystem.

Reviewer 2: ```suggestion
pods differently based on whether the node has a dedicated `imagefs` filesystem:
```

amount for each resource. When the kubelet notices that a resource is starved,
it continues to reclaim that resource until it reclaims the quantity you specify. 

For example, the following configuration sets minimum reclaim amounts: 

```
--eviction-hard=memory.available<500Mi,nodefs.available<1Gi,imagefs.available<100Gi
--eviction-minimum-reclaim="memory.available=0Mi,nodefs.available=500Mi,imagefs.available=2Gi"`
```

In this example, if the `nodefs.available` signal meets the eviction threshold,
the kubelet reclaims the resource until the signal reaches the threshold of `1Gi`,
and then continues to reclaim the minimum amount of `500Mi` it until the signal
reaches `1.5Gi`. 

Similarly, the kubelet reclaims the `imagefs` resource until the `imagefs.available`
signal reaches `102Gi`. 

The default `eviction-minimum-reclaim` is `0` for all resources.

## Scheduler

The scheduler respects reported node conditions when making scheduling decisions.

For example, if `MemoryPressure` node condition is active on a node, the scheduler
does not schedule new `BestEffort` pods on the node. If the `DiskPressure` node

> condition is active, the scheduler does not schedule any new pods on the node. 

Reviewer 2: This section is not really about eviction.

Reviewer 2: Maybe move it to https://kubernetes.io/docs/concepts/architecture/nodes/ or check that the content there already covers this?

immediately induce memory pressure.

Consider the following scenario:

* Node memory capacity: `10Gi`
* Operator wants to reserve 10% of memory capacity for system daemons (kernel, `kubelet`, etc.)
* Operator wants to evict Pods at 95% memory utilization to reduce incidence of system OOM.

For this to work, the kubelet is launched as follows:

```
--eviction-hard=memory.available<500Mi
--system-reserved=memory=1.5Gi
```

In this configuration, the `--system-reserved` flag reserves `1.5Gi` of memory
for the system, which is `10% of the total memory + the eviction threshold amount`. 

The node can reach the eviction threshold if a pod is using more than its request,
or if the system is using more than `1Gi` of memory, which makes the `memory.available`
signal fall below `500Mi` and triggers the threshold. 

### DaemonSet

Pod Priority is a major factor in making eviction decisions. If you do not want
the kubelet to evict pods that belong to a `DaemonSet`, give those pods a high
enough `priorityClass` in the pod spec. You can also use a lower `priorityClass`
or the default to only allow `DaemonSet` pods to run when there are enough 
resources.

> ## Deprecation of existing feature flags to reclaim disk 

Author: Is this still needed here?

Reviewer 2: I would check https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ and see if the “existing flag” names still exist.

To be honest, I think it's OK to omit this section even if the deprecated command line flags are still available. The page at https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ is the best place to find what command line flags are current.

`kubelet` has been freeing up disk space on demand to keep the node stable.

As disk based eviction matures, the following `kubelet` flags are marked for deprecation
in favor of the simpler configuration supported around eviction.

| Existing Flag                              | New Flag                                |
| ------------------------------------------ | ----------------------------------------|
| `--image-gc-high-threshold`                | `--eviction-hard` or `eviction-soft`    |
| `--image-gc-low-threshold`                 | `--eviction-minimum-reclaim`            |
| `--maximum-dead-containers`                | deprecated                              |
| `--maximum-dead-containers-per-container`  | deprecated                              |
| `--minimum-container-ttl-duration`         | deprecated                              |
| `--low-diskspace-threshold-mb`             | `--eviction-hard` or `eviction-soft`    |
| `--outofdisk-transition-frequency`         | `--eviction-pressure-transition-period` |

> ## Known issues

Author: Have these issues been fixed since this page was last updated?

Reviewer 2: I don't know who's best to confirm that. This advice does look stale. Try asking in `#sig-node` on the Kubernetes Slack.

---
title: Pod Eviction
content_type: concept
weight: 50
---

Eviction is the process by which pods in a node are gracefully terminated. 

In Kubernetes, there are two types of pod eviction:

  1. Manual eviction
  2. Out-of-resource eviction (kubelet eviction)

{{<caution>}}
Out-of-resource eviction does not share any characteristics
with manual eviction. Do not use the two eviction types interchangeably. 
{{</caution>}}

> ## Manual eviction

Reviewer: I'd suggest "Graceful Eviction", as that is the primary distinction.

Manual evictions respect your configured [`PodDisruptionBudgets`](/docs/tasks/run-application/configure-pdb/). 

## Out-of-resource eviction

{{< glossary_definition term_id="kubelet-eviction" length="all" >}}

During an out-of-resource eviction, the kubelet sets the `PodPhase` for the
selected pods to `Failed`. This terminates the pods. 

The kubelet does not respect your configured `PodDisruptionBudget` or the pod's
`terminationGracePeriodSeconds`, and uses a `0s` grace period for termination.

If the pods are managed by a workload resource (such as {{< glossary_tooltip text="StatefulSet" term_id="statefulset" >}}
or {{< glossary_tooltip text="Deployment" term_id="deployment" >}}) that
replaces failed pods, the control plane  creates new pods in place of the evicted pods.

{{<note>}}
The kubelet attempts to [reclaim node-level resources](#reclaim-node-resources)
before it terminates end-user pods. For example, it removes unused container
images when disk resources are starved.
{{</note>}}

The kubelet uses various parameters to make eviction decisions, like the following:

  * Eviction signals
  * Eviction thresholds
  * Monitoring intervals

>   * Node conditions

Reviewer: nit: node conditions are an _output_ of eviction, rather than an input

---
title: Pod Eviction
content_type: concept
weight: 50
---

Eviction is the process by which {{<glossary_tooltip text="pods" term_id="pod" length="short">}}
in a {{<glossary_tooltip text="node" term_id="node" length="short">}} are
terminated. 

In Kubernetes, there are two types of pod eviction:

  1. Graceful eviction
  2. Out-of-resource eviction ({{<glossary_tooltip text="kubelet" term_id="kubelet" length="short">}}
     eviction)

{{<caution>}}
Out-of-resource eviction does not share any characteristics
with graceful eviction. Do not use the two eviction types interchangeably. 
{{</caution>}}

## Graceful eviction

{{<glossary_definition term_id="graceful-eviction" length="all">}}

> Graceful evictions respect your configured [`PodDisruptionBudgets`](/docs/tasks/run-application/configure-pdb/). 

Reviewer: and also pod termination grace periods.

* Operator wants to evict Pods at 95% memory utilization to reduce incidence of system OOM.

For this to work, the kubelet is launched as follows:

```
--eviction-hard=memory.available<500Mi
--system-reserved=memory=1.5Gi
```

In this configuration, the `--system-reserved` flag reserves `1.5Gi` of memory
for the system, which is `10% of the total memory + the eviction threshold amount`. 

The node can reach the eviction threshold if a pod is using more than its request,
or if the system is using more than `1Gi` of memory, which makes the `memory.available`
signal fall below `500Mi` and triggers the threshold. 

#### DaemonSet

Pod Priority is a major factor in making eviction decisions. If you do not want
the kubelet to evict pods that belong to a `DaemonSet`, give those pods a high
enough `priorityClass` in the pod spec. You can also use a lower `priorityClass`
or the default to only allow `DaemonSet` pods to run when there are enough 
resources.

### Known issues

The following sections describe known issues related to out of resource handling.

#### kubelet may not observe memory pressure right away

> The `kubelet` currently polls `cAdvisor` to collect memory usage stats at a regular interval. If memory usage

Reviewer: We now have memcg notifications implemented, so change this to:

"By default, the `kubelet` polls ..."

---
title: Pod Eviction
content_type: concept
weight: 50
---

Eviction is the process by which {{<glossary_tooltip text="pods" term_id="pod" length="short">}}
in a {{<glossary_tooltip text="node" term_id="node" length="short">}} are
terminated. 

> In Kubernetes, there are two types of pod eviction:

Reviewer: nit: I don't think its the apiserver that controls graceful eviction.  Its probably some controller in the kube-controller-manager, but i'd double-check.

> In Kubernetes, there are two types of pod eviction:

Reviewer 2: > Graceful eviction is initiated by api-server

It's some _client_ of the API server (eg `kubectl drain`) that creates these Eviction subresources of Pod.

> In Kubernetes, there are two types of pod eviction:

Reviewer 2: The API client could also be a controller, whether that happens within the cluster's control plane or outside it.

> In Kubernetes, there are two types of pod eviction:

Author: so the Eviction object is created by interacting with the Eviction API through a client of apiserver like kubectl drain or a controller. When created, the apiserver (?) triggers graceful termination?

> In Kubernetes, there are two types of pod eviction:

Reviewer 3: ack, thanks for clarifying @dashpole and @stfim ; it's what I was trying to across - a client explicitly calls API server to create eviction object as opposed to kubelet out-of-resource eviction (which does not involve any external client creating eviction via API server).

Regarding what actually deletes pod during an eviction, I took a look and it appears it actually is api-server ([here](https://github.com/kubernetes/kubernetes/blob/2bda8ac3ec163ef158ca45d4f49de857c5107070/pkg/registry/core/pod/storage/eviction.go#L175-L180)). If the eviction cannot be performed due to outstading PDB it appears API-server will deny creation of the eviction object (and return `429` HTTP status) ([xref](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api)).

  1. Graceful eviction

>   2. Out-of-resource eviction ({{<glossary_tooltip text="kubelet" term_id="kubelet" length="short">}}

Reviewer 4: There are 2 types of out-of-resource eviction:

- The first is more like node resource-pressure eviction, acted on by the kubelet.
- The second is cluster level out-of-resource preemption, which clears space from any node to be able to schedule higher priority pods. This is acted on by scheduler.

Author: What's the difference between eviction and preemption? We have a separate topic talking about preemption. 

For this I can add: 
There are two types of out-of-resource eviction: 
* Node-level eviction: managed by the kubelet and triggered by resource pressure in nodes
* Cluster-level preemption: managed by the scheduler and triggered to clear space on nodes to schedule higher priority pods. For more information, see [Pod Priority and Preemption](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#preemption).

Reviewer 4: lgtm, but maybe @sftim has a better phrasing.

Author: @sftim - I updated the glossary entry for out-of-resource eviction. I'll work it into the doc in a way that makes sense and get you to look at it on my next push.

>   2. Out-of-resource eviction ({{<glossary_tooltip text="kubelet" term_id="kubelet" length="short">}}

Reviewer 2: ```suggestion
  2. Out-of-resource eviction ({{<glossary_tooltip text="kubelet" term_id="kubelet" >}}
```

     eviction)

> {{<caution>}}

Reviewer 2: This should be a `{{< note >}}` callout rather than a `{{< caution >}}`; we reserve `{{< caution >}}` for things that are a bit more serious (eg: potential for security problems, but only if you take specific steps). 

There is also `{{< warning >}}` for things you really ought to know (eg: don't use an allow-all authorization configuration except for local testing)

Author: Removed the note entirely.

Out-of-resource eviction does not share any characteristics

> with graceful eviction. Do not use the two eviction types interchangeably. 

Reviewer 4: You don't "use" out-of-resource eviction. The system applies it.

Author: I'll remove that sentence. Maybe "Out-of-resource eviction and graceful eviction are not interchangeable"?

Reviewer 4: I think the caution is unnecessary.

{{</caution>}}

## Graceful eviction

{{<glossary_definition term_id="graceful-eviction" length="all">}}

Graceful evictions respect your configured [`PodDisruptionBudgets`](/docs/tasks/run-application/configure-pdb/)
and [`terminationGracePeriodSeconds`](/docs/concepts/workloads/pods/pod-lifecycle#pod-termination). 

## Out-of-resource eviction

{{<glossary_definition term_id="kubelet-eviction" length="all">}}

During an out-of-resource eviction, the kubelet sets the `PodPhase` for the
selected pods to `Failed`. This terminates the pods. 

The kubelet does not respect your configured `PodDisruptionBudget` or the pod's
`terminationGracePeriodSeconds`, and uses a `0s` grace period for termination.

> If the pods are managed by a {{<glossary_tooltip text="workload" term_id="workload" length="short">}}

Reviewer 2: ```suggestion
If the pods are managed by a {{<glossary_tooltip text="workload" term_id="workload">}}
```
(`length` isn't relevant to tooltips; these either use the default text or the text you explicitly use as an override)

container, and if users use the [node
allocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable) feature, out of resource decisions
are made local to the end user Pod part of the cgroup hierarchy as well as the
root node. This [script](/examples/admin/resource/memory-available.sh)
reproduces the same set of steps that the kubelet performs to calculate
`memory.available`. The kubelet excludes inactive_file (i.e. # of bytes of
file-backed memory on inactive LRU list) from its calculation as it assumes that
memory is reclaimable under pressure.

The kubelet supports the following filesystem partitions:

1. `nodefs`: The node's root filesystem, used for local disk volumes, emptyDir,
   log storage, and more. For example, `nodefs` can provide `/` or `/var/lib/kubelet`. 
1. `imagefs`: An optional filesystem that container runtimes use to store container
   images and container writable layers.

Kubelet auto-discovers these filesystems and ignores other filesystems. Kubelet
does not support other configurations.

{{<note>}}
Some kubelet garbage collection features are deprecated in favor of eviction.

> For a list of the deprecated features, see [kubelet garbage collection deprecation](http://localhost:1313/docs/concepts/cluster-administration/kubelet-garbage-collection/#deprecation)

Reviewer 2: This link is incorrect - make it a site-relative URL.

To protect against oscillation, you can use the `eviction-pressure-transition-period`
flag, which controls how long the kubelet must wait before transitioning a node
condition to a different state. The transition period has a default value of `5m`.

### Reclaiming node level resources {#reclaim-node-resources}

The kubelet tries to reclaim node-level resources before it evicts end-user pods.

When a `DiskPressure` node condition is reported, the kubelet reclaims node-level
resources based on the filesystems on the node. 

#### With `imagefs`

If the node has a dedicated `imagefs` filesystem for container runtimes to use,
the kubelet does the following:

  * If the `nodefs` filesystem meets the eviction threshlds, the kubelet garbage collects
    dead pods and containers. 
  * If the `imagefs` filesystem meets the eviction thresholds, the kubelet
    deletes all unused images. 

#### Without `imagefs`

If the node only has a `nodefs` filesystem that meets eviction thresholds,
the kubelet frees up disk space in the following order:

1. Garbage collect dead pods and containers
1. Delete unused images

> ### Evicting end-user pods

Reviewer 2: ```suggestion
### Pod selection for kubelet eviction
```
?

that a resource is starved, it continues to reclaim that resource until it
reclaims the quantity you specify. 

For example, the following configuration sets minimum reclaim amounts: 

```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
evictionHard:
  memory.available: "500Mi"
  nodefs.available: "1Gi"
  imagefs.available: "100Gi"
evictionMinimumReclaim:
  memory.available: "0Mi"
  nodefs.available: "500Mi"
  imagefs.available: "2Gi"
```

In this example, if the `nodefs.available` signal meets the eviction threshold,
the kubelet reclaims the resource until the signal reaches the threshold of `1Gi`,
and then continues to reclaim the minimum amount of `500Mi` it until the signal
reaches `1.5Gi`. 

Similarly, the kubelet reclaims the `imagefs` resource until the `imagefs.available`
signal reaches `102Gi`. 

The default `eviction-minimum-reclaim` is `0` for all resources.

### Node out of memory behavior

If the node experiences an out of memory (OOM) event prior to the kubelet
being able to reclaim memory, the node depends on the [oom_killer](https://lwn.net/Articles/391222/)
to respond.

> The kubelet sets an `oom_score_adj` value for each container based on the QoS for the pod.

Reviewer 3: Might be worth to note somewhere that in addition to logic here if the pod has priority `system-node-critical` there is special case of it getting -997 score: 

https://github.com/kubernetes/kubernetes/blob/b58a7e2/pkg/kubelet/qos/policy.go#L41-L44

Reviewer 2: (that could come in a follow-up PR, BTW)

Author: Lol, yes, I am going to make new suggestions in new PRs (but I implemented this one already).

| Quality of Service | oom_score_adj                                                                     |
|--------------------|-----------------------------------------------------------------------------------|

> | `Guaranteed`       | -998                                                                              |

Reviewer 3: nit, but the code seems to set this to -997, not -998: https://github.com/kubernetes/kubernetes/blob/b58a7e2/pkg/kubelet/qos/policy.go#L30

it was changed in https://github.com/kubernetes/kubernetes/pull/71269

Author: So it's the same score as for system-node-critical priority pods?

Reviewer 3: yes, exactly: https://github.com/kubernetes/kubernetes/blob/b58a7e233e06f5e4b58a637af8e36cb4f59c001b/pkg/kubelet/qos/policy.go#L41-L44

---

> title: Pod Eviction

Reviewer 2: ```suggestion
title: Eviction
```
You can't evict (or schedule, or preempt, or deschedule) anything other than a pod.

content_type: concept
weight: 50
---

Eviction is the process by which {{<glossary_tooltip text="pods" term_id="pod">}}
in a {{<glossary_tooltip text="node" term_id="node">}} are
terminated. 

In Kubernetes, there are two types of pod eviction:

>   1. Graceful eviction

Reviewer 2: ```suggestion
  1. [Graceful eviction](#graceful-eviction)
  2. [Node-level eviction](#kubelet-eviction) ({{< glossary_tooltip text="kubelet" term_id="kubelet" >}}
     eviction)
```
_except_ that I'd also rename both terms; see feedback elsewhere.

  2. Node-level eviction ({{< glossary_tooltip text="kubelet" term_id="kubelet" >}}
     eviction)

## Graceful eviction {#graceful-eviction}

> {{< glossary_definition term_id="graceful-eviction" length="all" >}}

Reviewer 2: Maybe just use the short form, and then add extra text that's similar to the rest of the glossary definition.

Graceful evictions respect your configured [`PodDisruptionBudgets`](/docs/tasks/run-application/configure-pdb/)
and [`terminationGracePeriodSeconds`](/docs/concepts/workloads/pods/pod-lifecycle#pod-termination). 

## Node-level eviction {#kubelet-eviction}

> {{< glossary_definition term_id="kubelet-eviction" length="all" >}}

Reviewer 2: Maybe just use the short form, and then add extra text that's similar to the rest of the glossary definition.

> Pod eviction can also happen at the cluster level during [pod preemption](/docs/concepts/configuration/pod-priority-preemption/#preemption).

Reviewer 2: I disagree. Preemption is not eviction. We should instead clarify that preemption is a different mechanism by which the control plane stops a potentially healthy Pod.

Reviewer 4: +1, it should have it's own subtitle.

container, and if users use the [node
allocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable) feature, out of resource decisions
are made local to the end user Pod part of the cgroup hierarchy as well as the
root node. This [script](/examples/admin/resource/memory-available.sh)
reproduces the same set of steps that the kubelet performs to calculate
`memory.available`. The kubelet excludes inactive_file (i.e. # of bytes of
file-backed memory on inactive LRU list) from its calculation as it assumes that
memory is reclaimable under pressure.

The kubelet supports the following filesystem partitions:

1. `nodefs`: The node's main filesystem, used for local disk volumes, emptyDir,
   log storage, and more. For example, `nodefs` contains `/var/lib/kubelet/`. 
1. `imagefs`: An optional filesystem that container runtimes use to store container
   images and container writable layers.

Kubelet auto-discovers these filesystems and ignores other filesystems. Kubelet
does not support other configurations.

{{<note>}}
Some kubelet garbage collection features are deprecated in favor of eviction.

> For a list of the deprecated features, see [kubelet garbage collection deprecation](/docs/concepts/cluster-administration/kubelet-garbage-collection/#deprecation)

Reviewer 2: nit:
```suggestion
For a list of the deprecated features, see [kubelet garbage collection deprecation](/docs/concepts/cluster-administration/kubelet-garbage-collection/#deprecation).
```

|--------------------|-----------------------------------------------------------------------------------|
| `Guaranteed`       | -997                                                                              |
| `BestEffort`       | 1000                                                                              |
| `Burstable`        | min(max(2, 1000 - (1000 * memoryRequestBytes) / machineMemoryCapacityBytes), 999) |

{{<note>}}
The kubelet also sets an `oom_score_adj` value of `-997` for containers in Pods that have
`system-node-critical` {{<glossary_tooltip text="Priority" term_id="pod-priority">}}
{{</note>}}

If the kubelet can't reclaim memory before a node experiences OOM, the
`oom_killer` calculates an `oom_score` based on the percentage of memory it's
using on the node, and then adds the `oom_score_adj` to get an effective `oom_score`
for each container. It then kills the container with the highest score.

This means that containers in low QoS pods that consume a large amount of memory
relative to their scheduling requests are killed first.

Unlike pod eviction, if a container is OOM killed, the `kubelet` can restart it 
based on its `RestartPolicy`.

> ### Best practices

Reviewer 2: I'd be tempted to:
```suggestion
### Best practices {#kubelet-eviction-good-practice}
```

LRU list as the `active_file` statistic. The kubelet treats `active_file` memory
areas as not reclaimable. For workloads that make intensive use of block-backed 
local storage, including ephemeral local storage, kernel-level caches of file 
and block data means that many recently accessed cache pages are likely to be 
counted as `active_file`. If enough of these kernel block buffers are on the 
active LRU list, the kubelet is liable to observe this as high resource use and 
taint the node as experiencing memory pressure - triggering pod eviction.

For more more details, see [https://github.com/kubernetes/kubernetes/issues/43916](https://github.com/kubernetes/kubernetes/issues/43916)

You can work around that behavior by setting the memory limit and memory request
the same for containers likely to perform intensive I/O activity. You will need 
to estimate or measure an optimal memory limit value for that container.

## {{% heading "whatsnext" %}}

* Learn about [PodDisruptionBudgets](/docs/tasks/run-application/configure-pdb/)
* Learn about [Quality of Service](/docs/tasks/configure-pod-container/quality-service-pod/) (QoS)

> * Check out the [Eviction API](/docs/reference/generated/kubernetes-api/v1.21/#create-eviction-pod-v1-core)

Reviewer 2: We should set the version from code:
```suggestion
* Check out the [Eviction API](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#create-eviction-pod-v1-core)
```

---

> title: Eviction

Author: It's supposed to - it used to be Pod Eviction but as Tim said, only Pods can be evicted so I changed it to just Eviction :P

content_type: concept
weight: 50
---

Eviction is the process by which {{<glossary_tooltip text="pods" term_id="pod">}}
in a {{<glossary_tooltip text="node" term_id="node">}} are

> terminated. 

Author: We explain the nature of the termination in the details (and terminated sounds forceful).

> terminated. 

Author: We use _involuntary disruption_ to describe kubelet evictions in https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#voluntary-and-involuntary-disruptions, so maybe we call this "involuntarily terminate"?

> terminated. 

Reviewer 5: hmm, there are other types of involuntary disruption not mentioned here (like the VM itself getting evicted), I am fine keeping it as is.

> terminated. 

Reviewer 2: If feasible, [rewrite this phrase sentence in the active voice ](https://kubernetes.io/docs/contribute/style/style-guide/#use-active-voice)- this means we state what forces the termination.

> terminated. 

Author: Doesn't apply anymore after the refactoring, but you may have suggestions for the new glossary entry.

---
title: Pod Eviction
content_type: concept
weight: 50
---

*Eviction* is the process by which kubelet proactively fails pods to reclaim

> resources for nodes.

Reviewer 2: You could use a glossary definition shortcode here, if you make sure there is a suitable definition in the glossary.
https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#glossary

See https://github.com/kubernetes/website/blame/a22aa8fd4feec091e736bca12b2f8776ade0ac94/content/en/docs/concepts/overview/components.md#L17 for an example of where this is done elsewhere.

## How eviction works

> `kubelet` monitors resources like CPU, memory, disk space, and disk inodes on

Reviewer 2: BTW I'd write:
```suggestion
The kubelet monitors resources like CPU, memory, disk space, and filesystem inodes on
```

your cluster's nodes. When one or more of these resources reach specific
consumption levels, `kubelet` can proactively fail one or more pods on the node
to reclaim resources and prevent starvation. 

<!-- https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/ has many conceptual sections which I suggest we move to this topic. That task topic doesn't actually have any "tasks". -->

During an eviction, `kubelet` sets the `PodPhase` for the selected pods to `Failed`.
This gracefully terminates the pods. If the pods are managed by a Deployment, the
Deployment creates new pods for Kubernetes to schedule. 

> When evicting pods, `kubelet` respects your configured [PodDisruptionBudget](docs/tasks/run-application/configure-pdb/).
> You can control which pods are evicted last by setting the [QoS level](docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed)

Reviewer 2: ```suggestion
When evicting pods, `kubelet` respects your configured [PodDisruptionBudget](/docs/tasks/run-application/configure-pdb/).
You can control which pods are evicted last by setting the [QoS level](/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed)
```

on those pods to `Guaranteed`.

You can control the various parameters `kubelet` uses to make eviction decisions.

> For instructions, see [Configure Out of Resource Handling](docs/tasks/administer-cluster/out-of-resource/).

Reviewer 2: ```suggestion
For instructions, see [Configure Out of Resource Handling](/docs/tasks/administer-cluster/out-of-resource/).
```

### Eviction parameters

You can use various parameters to control eviction decisions, as follows: 

* Eviction thresholds
* Eviction signals
* Monitoring intervals
* 

## Triggering eviction

You can configure pod eviction to happen automatically, triggered by a controller. 

You can also request pod eviction manually using the `kubectl drain` command
or the [Eviction API](/docs/reference/generated/kubernetes-api/v1.21/#create-eviction-pod-v1-core). 

## What's next

> * Learn how to [Configure Out of Resource Handling](docs/tasks/administer-cluster/out-of-resource/)
> * Learn about [PodDisruptionBudgets](docs/tasks/run-application/configure-pdb/)
> * Learn about [Quality of Service (QoS)](docs/tasks/configure-pod-container/quality-service-pod/)

Reviewer 2: ```suggestion
* Learn how to [Configure Out of Resource Handling](/docs/tasks/administer-cluster/out-of-resource/)
* Learn about [PodDisruptionBudgets](/docs/tasks/run-application/configure-pdb/)
* Learn about [Quality of Service](/docs/tasks/configure-pod-container/quality-service-pod/) (QoS)
```
